rust/opendp/src/any.rs

```diff
@@ -0,0 +1,393 @@
+use std::any::Any;
+use std::cmp::Ordering;
+use std::rc::Rc;
+
+use crate::core::{Domain, Function, Measure, Measurement, Metric, PrivacyRelation, StabilityRelation, Transformation};
+use crate::error::*;
+use crate::traits::{FallibleSub, MeasureDistance, MetricDistance};
+
+pub trait Downcast {
+    fn downcast<T: 'static>(self) -> Fallible<T>;
+    fn downcast_ref<T: 'static>(&self) -> Fallible<&T>;
+}
+
+pub struct AnyBox {
+    value: Box<dyn Any>,
+    eq_glue: Rc<dyn Fn(&Self, &Self) -> bool>,
+}
+
+impl AnyBox {
+    pub fn new<T: 'static + PartialEq>(value: T) -> Self {
+        let eq_ = |self_: &Self, other: &Self| {
+            // This downcast of self will always succeed, so equality check is all that's necessary.
+            self_.value.downcast_ref::<T>() == other.value.downcast_ref::<T>()
+        };
+        Self {
+            value: Box::new(value),
+            eq_glue: Rc::new(eq_),
+        }
+    }
+}
+
+impl PartialEq for AnyBox {
+    fn eq(&self, other: &Self) -> bool {
+        (self.eq_glue)(self, other)
+    }
+}
+
+impl Downcast for AnyBox {
+    fn downcast<T: 'static>(self) -> Fallible<T> {
+        self.value.downcast().map_or_else(|_| fallible!(FailedCast), |v| *v)
+    }
+    fn downcast_ref<T: 'static>(&self) -> Fallible<&T> {
+        self.value.downcast_ref().ok_or_else(|| err!(FailedCast))
+    }
+}
+
+pub struct CloneableAnyBox {
+    value: Box<dyn Any>,
+    eq_glue: Rc<dyn Fn(&Self, &Self) -> bool>,
+    clone_glue: Rc<dyn Fn(&Self) -> Self>,
+}
+
+impl CloneableAnyBox {
+    pub fn new<T: 'static + PartialEq + Clone>(value: T) -> Self {
+        let eq_ = |self_: &Self, other: &Self| {
+            // This downcast of self will always succeed, so equality check is all that's necessary.
+            self_.value.downcast_ref::<T>() == other.value.downcast_ref::<T>()
+        };
+        let clone_ = |self_: &Self| {
+            Self {
+                value: Box::new(self_.value.downcast_ref::<T>().unwrap().clone()),
```

Comment:
Minor, but .unwrap_assert would be nice here. I don't consider this blocking.

Reply:
Oops good catch, I'll pick that up later.